types/node_definition.go

```diff
@@ -68,8 +68,8 @@ type NodeDefinition struct {
 	Sysctls map[string]string `yaml:"sysctls,omitempty"`
 	// Extra options, may be kind specific
 	Extras *Extras `yaml:"extras,omitempty"`
-	// List of node names to wait for before satarting this particular node
-	WaitFor []string `yaml:"wait-for,omitempty"`
+	// stage based dependency definition
```

Comment:
@steiler I want us to get rid of the

```go
Stage map[string]*WaitForEnvelop `yaml:",inline"`
```

and instead do a struct-based stage config

```go
// Stages represents a configuration of stages for a given node.
type Stages struct {
	Create      StageConfig `yaml:"create,omitempty"`
	CreateLinks StageConfig `yaml:"create-links,omitempty"`
	Configure   StageConfig `yaml:"configure,omitempty"`
	Healthy     StageConfig `yaml:"healthy,omitempty"`
	Exit        StageConfig `yaml:"exit,omitempty"`
}

// StageConfig represents a configuration of a given stage.
type StageConfig struct {
	WaitFor []*WaitFor `yaml:"wait-for,omitempty"`
}
```

We have a finite amount of stages, and doing a struct with fields feels more appropriate.
Unfortunately this means that changes to GetWaitFor are required, and I think it may be rather `GetStages()` as this is now the top level container that dependency manager should work on

Reply:
to identify one stage from the other we might create different types, like

```go
type CreateStageConfig struct {
  StageConfig
}
```
with an embedded common StageConfig. Then a String() method on those structs could return the string stage if required

Reply:
If we go down that path, we might even have seperate StageConfig structs.
Because There are certain options which will only be viable in certain stages.
So we could tage the generic StageConfig and implement that basic case. But include it in the specific structs, whcih would allow us to overwrite certain things per stage... kind of like the DefaultNode and Specific nodes implementations.
opinion?

Reply:
This seems to match what I added here https://github.com/srl-labs/containerlab/pull/1759#discussion_r1452051061 ?

Reply:
*Facepalm* ... yes! Sorry!